### PR TITLE
Integration of pr #37, unit tests and some bug fixes

### PR DESCRIPTION
**Pull request by NicoAlexH**

### CODE REVIEW

File: tests/attributes_testcase.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * ${PLUGINNAME} file description here.
+ *
+ * @package    ${PLUGINNAME}
+ * @copyright  2021 SysBind Ltd. <service@sysbind.co.il>
+ * @auther     alexandn
```
Comment (ndunand) [inline tests/attributes_testcase.php]: ?

File: edit.php [hunk 2]
```diff
@@ -44,6 +44,13 @@
     redirect($return);
 }
 
+/* No custom field defined */
+if(!$DB->get_records('user_info_field')){
+    $url = new moodle_url('/user/profile/index.php');
+    \core\notification::warning(get_string('no_custom_field', 'enrol_attributes', $url->get_scheme() .'://'. $url->get_host() . $url->get_path()));
```
Comment (ndunand) [inline edit.php]: just put '/user/profile/index.php' as 3rd arg

File: lang/fr/enrol_attributes.php [hunk 3]
```diff
@@ -16,5 +16,6 @@
 
 $string['group'] = 'Affectation aux groupes';
 $string['group_help'] = 'Vous pouvez affecter aucun ou plusieurs groupes';
+$string['no_custom_field'] = 'Il semblerait qu\'aucun attribut personnalisé ne soit défini. Cliquez <a href="{$a}" target="_blank">ici</a> pour en ajouter.';
```
Comment (ndunand) [inline lang/fr/enrol_attributes.php]: "Aucun attribut n'est défini"

File: settings.php [hunk 4]
```diff
@@ -58,8 +58,12 @@
         $settings->add(new admin_setting_configmultiselect('enrol_attributes/profilefields',
                 get_string('profilefields', 'enrol_attributes'), get_string('profilefields_desc', 'enrol_attributes'),
                 [], $customfields));
+    } elseif($PAGE->url->get_param('section') === 'enrolsettingsattributes'){
+        $url = new moodle_url('/user/profile/index.php');
+        \core\notification::warning(get_string('no_custom_field', 'enrol_attributes', $url->get_scheme() .'://'. $url->get_host() . $url->get_path()));
```
Comment (ndunand) [inline settings.php]: just put '/user/profile/index.php' as 3rd arg

File: lib.php [hunk 5]
```diff
@@ -51,8 +55,9 @@ public static function process_login(\core\event\user_loggedin $event) {
             $mapping = array();
             $mappings_str = explode("\n", str_replace("\r", '', get_config('enrol_attributes', 'mappings')));
             foreach ($mappings_str as $mapping_str) {
-                if (preg_match('/^\s*([^: ]+)\s*:\s*([^: ]+)\s*$/', $mapping_str, $matches) && in_array($matches[2],
-                                $customfields) && array_key_exists($matches[1], $_SERVER)) {
+                if (array_key_exists($matches[1], $_SERVER)
```
Comment (ndunand) [inline lib.php]: This can't work, no ? Using $matches before defining it.

File: lib.php [hunk 6]
```diff
@@ -213,18 +221,19 @@ public static function process_enrolments($event = null, $instanceid = null) {
...
-            }
-            else {
+            } else {
                 $users = $DB->get_records_sql($select . $arraysql['select'] . $where . $arraysql['where'],
                         $arraysql['params']);
                 $nbdbqueries++;
                 $cache->set($dbquerycachekey, serialize($users));
             }
-            foreach ($users as $user) {
+            foreach ($users ?? [] as $user) {
```
Comment (ndunand) [inline lib.php:234]: ok in PHP 7.2 ?

Comment (ndunand) [reply lib.php:234]: (since we support Moodle 3.5 (I think) we therefore also must support old PHP versions

File: lib.php [hunk 7]
```diff
@@ -306,26 +315,26 @@ public static function arraysyntax_tosql($arraysyntax, &$join_id = 0) {
...
+            } elseif ($customkey = array_search($rule->param, $customuserfields, true)) {
+                // custom user field actually exists
+                $join_id++;
+                $data = 'd' . $join_id . '.data';
+                $select .= ' RIGHT JOIN {user_info_data} d' . $join_id . ' ON d' . $join_id . '.userid = u.id AND d' . $join_id . '.fieldid = ' . $customkey;
+                $where .= ' (' . $DB->sql_compare_text($data) . ' = ' . $DB->sql_compare_text('?') . ' OR ' . $DB->sql_like($DB->sql_compare_text($data),
+                                '?') . ' OR ' . $DB->sql_like($DB->sql_compare_text($data),
+                                '?') . ' OR ' . $DB->sql_like($DB->sql_compare_text($data), '?') . ')';
+                array_push($params, $rule->value, '%;' . $rule->value, $rule->value . ';%',
+                        '%;' . $rule->value . ';%');
```
Comment (ndunand) [inline lib.php:331]: for readability, this should really go into a single "code reformating" single commit

File: tests/attributes_testcase.php (hunk 1 above)
Comment (NicoAlexH) [reply tests/attributes_testcase.php]: copy / paste fail

File: edit.php (hunk 2 above)
Comment (NicoAlexH) [reply edit.php]: done

File: lang/fr/enrol_attributes.php (hunk 3 above)
Comment (NicoAlexH) [reply lang/fr/enrol_attributes.php]: done

File: lib.php (hunk 5 above)
Comment (NicoAlexH) [reply lib.php]: Fixed

File: lib.php (hunk 6 above)
Comment (NicoAlexH) [reply lib.php:234]: Yep, null coalescing operator is supported from PHP 7.0 onwards : https://www.php.net/manual/en/migration70.new-features.php

File: lib.php (hunk 7 above)
Comment (NicoAlexH) [reply lib.php:331]: Got it ! sorry :)

**Comment by ndunand** [comment]
Merged! thanks @NicoAlexH and @DeleauRomain

**Pull request merged**